framework/src/org/checkerframework/framework/type/BoundsInitializer.java

```diff
@@ -391,36 +465,76 @@ public void initializeExtendsBound(final AnnotatedWildcardType wildcard) {
-                        final AnnotatedTypeMirror annoTypeArg =
-                                AnnotatedTypeMirror.createType(
-                                        actualTypeArgs.get(i), declaredType.atypeFactory, false);
-
-                        final BoundPathNode node = addPathNode(new TypeArgNode(i));
-                        typeArgs.add(replaceOrVisit(annoTypeArg));
-                        removePathNode(node);
+                typeArgs = new ArrayList<>();
+                for (int i = 0; i < typeElement.getTypeParameters().size(); i++) {
+                    TypeMirror javaTypeArg =
+                            getJavaType(declaredType, typeElement.getTypeParameters(), i);
+                    final AnnotatedTypeMirror atmArg =
+                            AnnotatedTypeMirror.createType(
+                                    javaTypeArg, declaredType.atypeFactory, false);
+                    typeArgs.add(atmArg);
+                    if (atmArg.getKind() == TypeKind.WILDCARD && declaredType.wasRaw()) {
+                        ((AnnotatedWildcardType) atmArg).setUninferredTypeArgument();
                     }
                 }
-                declaredType.setTypeArguments(typeArgs);
             } else {
+                typeArgs = declaredType.typeArgs;
+            }
 
-                final List<AnnotatedTypeMirror> typeArgs =
-                        new ArrayList<>(declaredType.getTypeArguments());
-                final List<AnnotatedTypeMirror> typeArgReplacements =
-                        new ArrayList<>(typeArgs.size());
-                for (int i = 0; i < typeArgs.size(); i++) {
-                    final AnnotatedTypeMirror typeArg = typeArgs.get(i);
-                    final BoundPathNode node = addPathNode(new TypeArgNode(i));
-                    typeArgReplacements.add(replaceOrVisit(typeArg));
-                    removePathNode(node);
+            List<AnnotatedTypeMirror> typeArgReplacements = new ArrayList<>(typeArgs.size());
+            for (int i = 0; i < typeArgs.size(); i++) {
+                AnnotatedTypeMirror typeArg = typeArgs.get(i);
+                BoundPathNode node = addPathNode(new TypeArgNode(i));
+                if (typeArg.getKind() == TypeKind.WILDCARD) {
+                    ((AnnotatedWildcardType) typeArg)
+                            .setTypeVariable(typeElement.getTypeParameters().get(i));
                 }
+                typeArgReplacements.add(replaceOrVisit(typeArg));
+                removePathNode(node);
+            }
+
+            declaredType.setTypeArguments(typeArgReplacements);
+        }
 
-                declaredType.setTypeArguments(typeArgReplacements);
+        /**
+         * Store the wildcards created as type arguments to raw types. This way if the raw type is
+         * visited again, the same wildcard can be used.
+         */
+        private Map<TypeVariable, WildcardType> rawTypeWildcards = new HashMap<>();
```

Comment:
Can the field be final?
When can the map be cleared? Or should it be a LRU map to limit its size?

Reply:
I'll make it final.  There's no need to limit or clear the cache as a new `InitializerVisitor` is created for each type that is initialized.